equational-reasoning.md

```diff
@@ -0,0 +1,25 @@
+Equational reasoning
+====================
+
+**Tags**: *substitution model*, *referential transparency*, *purity*, *side effect*
+
+In terms of programming we can say that a function definition defines the scope where different expressions are evaluated in order to eventually return a result. This result defines the return type of that function.
+
+For developers it seems very convenient to be able to `reason` about a function (i.e. understand what the function does ) just by checking the value that it returns. This is to say that a function's return value represents the transformation of its internal state by potentially using other arguments as its parameters.
+
+However what a function does is not always related to the value that it returns and this is the case when a function has side effects. This is when we say that a function is not `pure`. By side effect we mean any kind of interaction that a function has with the outside word (i.e. outside of its scope ). This can be IO operations, altering state of external variables etc. As a result, the only way to reason about what the function does is to deep dive into the definition of the function and monitor what happens when a side effect operation happens.
```

Comment:
`As a result, the only way to reason about what the function does is to deep dive into the definition of the function and monitor what happens when a side effect operation happens.` <- I like this a lot

Reply:
It is not clear to me why I should monitor what happens in order to understand. When I see `println` or `gettimeinmillis` I don't need to monitor in order to reason.

Reply:
one thing i can think of is starting a new thread that modifies global state

Reply:
Yes, simple functionas like `println` are simple to understand, but compose side effect functions and you will need to deep dive into the definition of each sub function to be able to know everything that it does, in contrast, a pure function, no matter how complex and how many sub expressions it may have, you only need to think about the input/output to reason about it

Reply:
my objection is against having to `monitor what happens`, not against having to examine the implementation.